hadoop-hdds/docs/content/design/degraded-storage-volumes.md

```diff
@@ -0,0 +1,275 @@
+
+#### IO Error Count
+
+This would be a new check that can be used as part of this feature. Currently each time datanode IO encounters an error, we request an on-demand volume scan. This should include every time the container scanner marks a container unhealthy. We can keep a counter of how many IO errors have been reported on a volume over a given time frame, regardless of whether the corresponding volume scan passed or failed. This accounts for cases that may show up on the main IO path but may otherwise not be detected by the volume scanner. For example, numerous old sectors with existing container data may be unreadable. The volume scanner's **File Check** will only utilize new disk sectors so it will still pass with these errors present, but the container scanner may be hitting many bad sectors across containers, which this check will account for.
+
+#### Sliding Window
+
+Most checks will encounter intermittent issues, even on overall healthy drives, so we should not downgrade volume health state after just one error. The current volume scanner uses a counter based sliding window for intermittent failues, meaning the volume will be failed if `x` out of the last `y` checks failed, regardless of when they occurred. This approach works for background volume scans, because `y` is the number of times the check ran, and `x` is the number of times it failed. It does not work if we want to apply a sliding window to on-demand checks like IO error count that do not care if the corresponding volume scan passed or failed.
+To handle this, we can switch to time based sliding windows to determine when a threshold of tolerable errors is crossed. For example, if this check has failed `x` times in the last `y` minutes, we should consider the volume degraded.
+
+We can use one time based sliding window to track errors that would cause a volume to be degraded, and a second one for errors that would cause a volume to be failed. When a check fails, it can add the result to whichever sliding window it corresponds to. We can create the following assignments of checks:
+
+- **Directory Check**: No sliding window required. If the volume is not present based on filesystem metadata it should be failed immediately.
+- **Database Check**: On failure, add an entry to the **failed health sliding window**
+- **File Check**:
+    - If the sync portion of the check fails, add an entry to the **failed health sliding window**
+    - If any other part of this check fails, add an entry to the **degraded health sliding window**
+- **IO Error Count**: When an on-demand volume scan is requested, add an entry to the **degraded health sliding window**
```

Comment:
This is how the container scanner informs the volume scanner of a problem, correct? 
Why do we need a sliding window? If the total of X I/O errors was reported, decide to fail it. Sliding window to me makes a decision to prioritize errors based on time, but then it is complicated to implement, instead, a threshold is a simple measure. Even in this case, how do you decide what is X? What heuristic guides this decision?
Also, if X cannot be decided very easily, ignoring the X - length(window) events seems even more aspirational.